gh-113977, gh-120754: Remove unbounded reads from zipfile

cmaloney opened the pull request:
Read without a size may read an unbounded amount of data + allocate unbounded size buffers. Move to capped size reads to prevent potential issues.

This is a followup https://github.com/python/cpython/pull/121593#issuecomment-2222261986. I don't think this needs a separate news entry.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNN: Summary of the changes made
```

Where: gh-NNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `main`.

-->


<!-- gh-issue-number: gh-120754 -->
* Issue: gh-120754
<!-- /gh-issue-number -->


<!-- gh-issue-number: gh-113977 -->
* Issue: gh-113977
<!-- /gh-issue-number -->

cmaloney committed:
GH-120754: Remove unbounded reads from zipfile

Read without a size may read an unbounded amount of data + allocate
unbounded size buffers. Move to capped size reads to prevent potential
issues.

picnixz commented on Lib/zipfile/__init__.py:
```diff
@@ -331,7 +331,7 @@ def _EndRecData(fpin):
     maxCommentStart = max(filesize - (1 << 16) - sizeEndCentDir, 0)
     fpin.seek(maxCommentStart, 0)
-    data = fpin.read()
+    data = fpin.read(ZIP_MAX_COMMENT + sizeEndCentDir + 1)
```
Maybe add an assert checking that you are indeed at the end of the file? (just in case).

cmaloney replied:
I don't see a good way to test that locally. In theory the seek just before this calculated that out and moved so that this read size will always read past the end. `test_zipfile` does have a test for appending a comment which exercises this

danifus commented on Lib/zipfile/__init__.py:
```diff
@@ -331,7 +331,7 @@ def _EndRecData(fpin):
     # number does not appear in the comment.
     maxCommentStart = max(filesize - (1 << 16) - sizeEndCentDir, 0)
```
While not part of your changes, I think this line might be wrong. It should allow up to `ZIP_MAX_COMMENT = (1 << 16) - 1` bytes for the comment (one less than currently)

```suggestion
    maxCommentStart = max(filesize - ZIP_MAX_COMMENT - sizeEndCentDir, 0)
```

cmaloney replied:
Was wondering about that. Does it makes sense to make a variable for `ZIP_MAX_COMMENT + sizeEndCentDir` variable for both spots to use? (if the `fpin.seek` jus below can be changed to a SEEK_END like the previous seek, this code also no longer needs to do its own `filesize` fetching + calculation)

danifus replied:
You could make a variable for `ZIP_MAX_COMMENT + sizeEndCentDir` but I'm happy either way.

With regards to the `filesize` variable, you could rewrite it to avoid that first `fpin.seek(0, 2)` but we still need to have the absolute location of the start of the central directory as we use it for this line in `_RealGetContents`:
`        self.start_dir = offset_cd + concat`
So you could get the (potential) central directory offset directly using a `tell()` after
```
    try:
        fpin.seek(-sizeEndCentDir, 2)
    except OSError:
        return None
```
succeeds and do the rest of the offset calculations relative to that. That would be one less `seek()`. Probably best to do that in a different PR if you want

cmaloney replied:
Will leave as is for now / focused on other bits around I/O stack

danifus commented on Lib/zipfile/__init__.py:
```diff
@@ -331,7 +331,7 @@ def _EndRecData(fpin):
     maxCommentStart = max(filesize - (1 << 16) - sizeEndCentDir, 0)
     fpin.seek(maxCommentStart, 0)
-    data = fpin.read()
+    data = fpin.read(ZIP_MAX_COMMENT + sizeEndCentDir + 1)
```
With the change to line 332 you can remove the `+ 1` and `test.test_zipfile.test_core.OtherTests.test_comments` will still pass.

```suggestion
    data = fpin.read(ZIP_MAX_COMMENT + sizeEndCentDir)
```

If you don't want to make the change to line 332 (as it is a little off topic), I think we should be consistent with how the number of bytes is calculated in line 332:
```suggestion
    data = fpin.read((1 << 16) + sizeEndCentDir)
```

I like how you've done it and my preference would be to fix line 332

danifus commented on Lib/zipfile/__init__.py:
```diff
@@ -309,7 +309,7 @@ def _EndRecData(fpin):
     except OSError:
         return None
-    data = fpin.read()
+    data = fpin.read(sizeEndCentDir + 1)
```
We've just seeked back `sizeEndCentDir` bytes so should be ok reading the same number of bytes

```suggestion
    data = fpin.read(sizeEndCentDir)
```

cmaloney replied:
There is a slight behavior change, which may entirely be okay, if don't do the `+1` compared to the `read()` without a size (`readall`). The extra byte is needed so `read()` does a second `os.read()` to try getting that byte and either gets more data than expected (someone wrote to the file while this was reading / size changed unexpectedly), or the second `os.read()` returns size 0 which indicates "this is EOF" (how `readall` decides to stop). The extra byte case it felt like was trying to be tested for just below in the `len(data) == sizeEndCentDir` case.

Happy to defer to others judgment around what zipfile should do in this case, it's reasonable to me to not read to EOF.

cmaloney replied:
`_Py_read` not `os.read` actually (os.read is a whole different codepath)

danifus replied:
I don't think it is required but I'm also happy to defer to others.

Here's the issue / PR that added the length check for context:
https://bugs.python.org/issue4844
https://github.com/python/cpython/commit/d2b1527f1418f4bac9d05933623ef1a9d1948f61

That check may have been added along with other similar checks across this file but not actually necessary due to the try except block that precedes it.

danifus replied:
If we do end up keeping the `+ 1`, we should add a comment detailing why it is there.

cmaloney committed:
Update Lib/zipfile/__init__.py

Co-authored-by: Daniel Hillier <daniel.hillier@gmail.com>

cmaloney committed:
Update Lib/zipfile/__init__.py

Co-authored-by: Daniel Hillier <daniel.hillier@gmail.com>

cmaloney committed:
Don't require reading an EOF

The seek just before this _should_ result in this read hitting EOF here or getting less bytes.

Co-authored-by: Daniel Hillier <daniel.hillier@gmail.com>

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

cmaloney committed:
Merge branch 'main' into cmaloney/zipfile_tweaks
